### PR TITLE
Asserter generator

### PR DESCRIPTION
**Pull request by agallou**
new version of the generator asserter after the discussions on #661

it's a work in progress, the unit tests are not fixed, but the functional test works.

@Hywan @mikaelrandy @jubianchi what are your thoughts on that ?

with the code on the PR this test passes : 

``` php
    public function testUsage()
    {
        $generator = function() {
            for ($i=0; $i<3; $i++) {
                yield ($i+1);
            }

            return 42;
        };

        $this
            ->generator($generator())
                ->yields->variable->isEqualTo(1)
                ->yields->variable->isEqualTo(2)
                ->yields->variable->isEqualTo(3)
                ->yields->variable->isNull()
                ->returns->isEqualTo(42)
            ->generator($generator())
                ->size->isEqualTo(3)
        ;
    }

    public function testUsageComplect()
    {
        $generator = function() {
            yield [
                "1", "2", "3"
            ];

            yield 0;
            yield 0;

            yield ["a" => 1, "b" => 2, "c" => 3];

            return 42;
        };

        $this
            ->generator($generator())
                ->yields
                    ->array
                        ->isEqualTo(["1", "2", "3"])
                        ->contains("1")
                        ->string[0]->isEqualTo(1)
                ->yields
                    ->integer
                        ->isZero()
                ->yields
                    ->integer
                        ->isZero
                ->yields
                    ->array
                        ->keys->isEqualTo(['a', 'b', 'c'])
                ->yields
                    ->variable->isNull
                ->returns->isEqualTo(42)
        ;
    }
```

### CODE REVIEW

File: tests/functionals/classes/asserters/generator.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+class generator extends atoum\tests\functionals\test\functional
+{
+	/**
+	 * @php >= 7.0
+	 */
+	public function testUsage()
+	{
+		if (version_compare(PHP_VERSION, '7.0') >= 0) {
+			$generator = eval(<<<'PHP'
```
Comment (Hywan) [inline tests/functionals/classes/asserters/generator.php]: Why an eval? In case PHP does not support the syntax?

**Comment by Hywan** [comment]
Much better for the user API! Great work :-]! :+1:

File: tests/functionals/classes/asserters/generator.php (hunk 1 above)
Comment (agallou) [inline tests/functionals/classes/asserters/generator.php]: yes, in php 5.6 there is an error when the generator returns something, even with the version_compare. The eval avoids that.

**Comment by agallou** [comment]
With the lastest version, the any asserter could by used after the returns property. 

For example : 

```php
    public function testUsage()
    {
        $generator = function() {
            yield [
                "1", "2", "3"
            ];

            yield 0;
            yield 0;

            yield ["a" => 1, "b" => 2, "c" => 3];

            return 42;
        };

        $this
            ->generator($generator())
                ->yields->array
                      ->isEqualTo(["1", "2", "3"])
                      ->contains("1")
                      ->string[0]->isEqualTo(1)
                ->yields->integer
                      ->isZero()
                ->yields->integer
                      ->isZero
                ->yields->array
                      ->keys->isEqualTo(['a', 'b', 'c'])
                ->yields->variable
                      ->isNull
                ->returns->variable
                      ->isEqualTo(42)
        ;
```

**Comment by agallou** [comment]
PR rebased from the master to remove conflicts, code style fixed and commits squashed. IMHO it's ready to be merged.

**Comment by Hywan** [comment]
Perfect!

**Comment by jubianchi** [comment]
Good job!

Thanks @agallou! 

👍 🎉

**Pull request merged**